### PR TITLE
Easier way to customize templates.

### PR DESCRIPTION
**Pull request by JaroslavTulach**
Ever annoyed by _To change this license header, choose License Headers in Project Properties._ and _To change this template file, choose Tools | Templates_ let's replace that with a clickable hyperlink:
image
The `nbfs:` URL would be recognized as a hyperlink and a click would open the template file in the editor.

### CODE REVIEW

**Review by sdedic — APPROVED** [review]
Perfect, although I don't like that `nbfs:` sufacing right out to the user. Sadly, the Fold APIs do not support actionable folds (they already support content preview) - so [NETBEANS-5684](https://issues.apache.org/jira/browse/NETBEANS-5684).

**Comment by JaroslavTulach** [comment]
> I don't like that `nbfs:` sufacing right out to the user. 

One of my earlier prototypes was using `config:/Templates/Classes/Class.java` - however then I decided to go public without introducing new protocol. That decision can change as this is really very visible, user facing feature.

**Comment by junichi11** [comment]
URL is not highlighted.

#### Before:
netbeans-pr-2960-before


#### After:
netbeans-pr-2960-after

#### Exception:
```
java.lang.UnsupportedOperationException: Not supported yet.
	at org.netbeans.modules.editor.url.HighlightURLs$3.subSequence(HighlightURLs.java:217)
	at java.util.regex.Matcher.getSubSequence(Matcher.java:1294)
	at java.util.regex.Matcher.group(Matcher.java:541)
	at org.netbeans.modules.editor.url.Parser.recognizeURLsREBased(Parser.java:44)
	at org.netbeans.modules.editor.url.Parser.recognizeURLs(Parser.java:36)
	at org.netbeans.modules.editor.url.HighlightURLs.run(HighlightURLs.java:222)
	at org.openide.util.RequestProcessor$Task.run(RequestProcessor.java:1418)
	at org.netbeans.modules.openide.util.GlobalLookup.execute(GlobalLookup.java:45)
	at org.openide.util.lookup.Lookups.executeWith(Lookups.java:278)
	at org.openide.util.RequestProcessor$Processor.run(RequestProcessor.java:2033)
Caused: org.openide.util.RequestProcessor$FastItem: task failed due to
```

Maybe `recognizeURLs()` is called many times, I suppose. Is the performance problem OK although I can be wrong?
(about using `URL_PATTERN.matcher(text)` instead)

**Comment by JaroslavTulach** [comment]
With the most recent change I can open the `nbfs:` URLs even in VSCode:

obrazek

File: enterprise/micronaut/src/org/netbeans/modules/micronaut/hyperlink/HyperlinkImpl.java [hunk 1]
```diff
@@ -0,0 +1,290 @@
+/*
```
Comment (dbalek) [inline enterprise/micronaut/src/org/netbeans/modules/micronaut/hyperlink/HyperlinkImpl.java:1]: I do not like location of this file inside of `enterprise/micronaut` module. I would prefer leaving this file as a part of `ide/editor` module and adding its dependency on `ide/api.lsp` if necessary.

File: enterprise/micronaut/src/org/netbeans/modules/micronaut/hyperlink/HyperlinkImpl.java [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+                } catch (IOException ex) {
+                    Exceptions.printStackTrace(ex);
+                }
+            }
+        }
+
+        return null;
+    }
+    
+    @MimeRegistration(mimeType = "text/x-java", service = HyperlinkLocationProvider.class)
```
Comment (dbalek) [inline enterprise/micronaut/src/org/netbeans/modules/micronaut/hyperlink/HyperlinkImpl.java:270]: Try to use empty mime path to register this provider for all files.

**Review by dbalek — APPROVED** [review]
Very useful feature. Thanks.

**Comment by JaroslavTulach** [comment]
There is a way to [registerTextDocumentContentProvider](https://code.visualstudio.com/api/references/vscode-api#workspace.registerTextDocumentContentProvider) - probably it should be used to handle the `nbfs` protocol and delay materialization of the file on disk until really needed - thanks for finding this out, @dbalek!

**Comment by JaroslavTulach** [comment]
OK, @junichi11 - RegExp in Java is known to be quite slow:

> Test Result: times=2049 vs. 19441

As such I am keeping the old manually written parser: ce7a42a - using the old parser should also fix the problem with 

> URL is not highlighted.

it seems, it is highlighted now.

File: enterprise/micronaut/src/org/netbeans/modules/micronaut/hyperlink/HyperlinkImpl.java (hunk 2 above)
Comment (JaroslavTulach) [reply enterprise/micronaut/src/org/netbeans/modules/micronaut/hyperlink/HyperlinkImpl.java]: That didn't work, but hopefully we won't need this part at all.

File: enterprise/micronaut/src/org/netbeans/modules/micronaut/hyperlink/HyperlinkImpl.java (hunk 1 above)
Comment (JaroslavTulach) [reply enterprise/micronaut/src/org/netbeans/modules/micronaut/hyperlink/HyperlinkImpl.java]: Removed in ee9eb43

File: enterprise/micronaut/src/org/netbeans/modules/micronaut/hyperlink/Parser.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
+/*
```
[the remark this answers is not in the document]
Comment (JaroslavTulach) [reply enterprise/micronaut/src/org/netbeans/modules/micronaut/hyperlink/Parser.java]: Done. `editor` now depends on `lsp.api`.

**Review by junichi11 — APPROVED** [review]
Perfect!
I've verified 
- there is no exception
- paths are highlighted
- jump to templates when I ([Ctrl]+)click paths

Thank you!  
nb-pr-2960

**Comment by JaroslavTulach** [comment]
In addition to here-in provided work I can also offer [delegating hyperlink provider](https://github.com/apache/netbeans/commit/a07f2f8698b462b728cbf976bb5398d16e434f6b) that can bridge `lsp.api` and NetBeans editor. However review by @dbalek suggested to keep that change out of this PR. Including here for completeness. The code may be handy in the future.

**Pull request merged**